### PR TITLE
perf(fuse): add/remove "used-by-pod" annotations on fuse pod when publishing/unpublishing

### PR DESCRIPTION
**Pull request by AlbeeSo**
#### What this PR does / why we need it:
add/remove annotations like `used-by-pod/09ec7b80-71d1-4fe9-8fa1-ea234563b05a: "20240116195325"` on fuse pod when publishing/unpublishing, to help users understand the mapping between application pods and fuse pods
#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note

```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: pkg/oss/utils.go [hunk 1]
```diff
@@ -99,6 +99,16 @@ func parseOtherOpts(otherOpts string) (mountOptions []string, err error) {
 	return mountOptions, nil
 }
 
+func getPodUid(targetPath, volumeID string) (podUid string) {
```
Comment (huww98) [inline pkg/oss/utils.go:102]: See https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.28/#csidriverspec-v1-storage-k8s-io (podInfoOnMount) for how to get pod info. Maybe we should not rely on the format of targetPath.

File: pkg/mounter/fuse_containerized_mounter.go [hunk 2]
```diff
@@ -273,9 +292,9 @@ func (mounter *ContainerizedFuseMounter) launchFusePod(ctx context.Context, sour
 		var rawPod corev1.Pod
 		rawPod.GenerateName = fmt.Sprintf("csi-fuse-%s-", mounter.name())
 		rawPod.Labels = labels
-		rawPod.Annotations = map[string]string{FuseMountPathAnnoKey: target, FuseSafeToEvictAnnoKey: "true"}
+		rawPod.Annotations = map[string]string{FuseMountPathAnnoKey: target, FuseSafeToEvictAnnoKey: "true", getUsedByPodAnnoKey(mounter.podUid): time.Now().Format("20060102150405")}
```
Comment (huww98) [inline pkg/mounter/fuse_containerized_mounter.go:295]: Why not standard format? e.g. `time.RFC3339`

**Review by huww98 — COMMENTED** [review]
I would recommend `csi.alibabacloud.com/used-by-pod/namespace/pod-name` as annotation name.

File: pkg/oss/utils.go (hunk 1 above)
Comment (AlbeeSo) [reply pkg/oss/utils.go:102]: it seems that only NodePublishVolumeRequest  has related volume_context of podInfoOnMount. But we need it on unpublish stage.
BTW, `csi.alibabacloud.com/used-by-pod/namespace/pod-name` is invalid. annotation key could only has ONE "/".

**Comment by huww98** [comment]
This annotation should always have the same result as `kubectl describe pvc xxx` now, right?

But if this is combined with #957 , consider:
1. Pod "a" start, fuse pod "fuse-a" start with used-by-pod/a
2. fuse pod "fuse-a" fail
3. Pod "b" start, fuse pod "fuse-b" start with used-by-pod/b
4. Pod "b" finished, then "fuse-b" is left with no used-by-pod

This seems OK.

File: pkg/oss/utils.go (hunk 1 above)
Comment (huww98) [reply pkg/oss/utils.go:102]: OK, then the part before "/" should be a domain name.

**Comment by AlbeeSo** [comment]
> This annotation should always have the same result as `kubectl describe pvc xxx` now, right?
> 
> But if this is combined with #957 , consider:
> 
> 1. Pod "a" start, fuse pod "fuse-a" start with used-by-pod/a
> 2. fuse pod "fuse-a" fail
> 3. Pod "b" start, fuse pod "fuse-b" start with used-by-pod/b
> 4. Pod "b" finished, then "fuse-b" is left with no used-by-pod
> 
> This seems OK.

here is a small demo for fuse pod crash case:

1. create  an app pod with oss pv, with a fuse pod `csi.alibabacloud.com/used-by-pod-5c7af695-e754-4b5b-9f08-609a76f6eac2: "2024-01-18T20:45:02+08:00"`
2. fuse pod crash by killing fuse daemon
3. create another app pod, fuse restart with annotation `csi.alibabacloud.com/used-by-pod-c0e4839c-8f73-41de-99ad-e998cddbdaba: "2024-01-18T21:01:09+08:00"`
4. pod `5c7af695-e754-4b5b-9f08-609a76f6eac2` need to restart even it is on pvc's used-by list.

**Comment by huww98** [comment]
I still have concerns about using UID in the user interface.

How about:
* label: csi.alibabacloud.com/namespace: "<namespace of pvc/pod>"
* annotation: csi.alibabacloud.com/used-by-pod: "\<pod-name1\>,\<pod-name2\>"

I know this is harder to implement, as we need to pull the old annotation to add new pod. But it may be more friendly to users.

`csi.alibabacloud.com/used-by-pod-<pod name>` have issue about the length limit on annotation name

**Comment by AlbeeSo** [comment]
> I still have concerns about using UID in the user interface.
> 
> How about:
> 
> * label: csi.alibabacloud.com/namespace: "<namespace of pvc/pod>"
> * annotation: csi.alibabacloud.com/used-by-pod: "<pod-name1>,<pod-name2>"
> 
> I know this is harder to implement, as we need to pull the old annotation to add new pod. But it may be more friendly to users.
> 
> `csi.alibabacloud.com/used-by-pod-<pod name>` have issue about the length limit on annotation name

the point is we cannot get the pod name on unpublish stage unless we record it.

**Pull request closed without merging**